perception/shape_estimation/lib/model/boost_bounding_box.cpp

```diff
@@ -0,0 +1,54 @@
+// Copyright 2018 Autoware Foundation. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "shape_estimation/model/boost_bounding_box.hpp"
+
+#include <boost/math/tools/minima.hpp>
+
+constexpr float epsilon = 0.001;
+
+BoostBoundingBoxShapeModel::BoostBoundingBoxShapeModel() : BoundingBoxShapeModel() {}
+
+BoostBoundingBoxShapeModel::BoostBoundingBoxShapeModel(
+  const boost::optional<ReferenceYawInfo> & ref_yaw_info)
+: BoundingBoxShapeModel(ref_yaw_info)
+{
+}
+
+float BoostBoundingBoxShapeModel::optimize(
+  const pcl::PointCloud<pcl::PointXYZ> & cluster, const float min_angle, const float max_angle)
+{
+  // std::cerr << "Boost ! " << std::endl;
+  auto closeness_func = [&](float theta) {
+    Eigen::Vector2f e_1;
+    e_1 << std::cos(theta), std::sin(theta);  // col.3, Algo.2
+    Eigen::Vector2f e_2;
+    e_2 << -std::sin(theta), std::cos(theta);  // col.4, Algo.2
+    std::vector<float> C_1;                    // col.5, Algo.2
+    std::vector<float> C_2;                    // col.6, Algo.2
+    for (const auto & point : cluster) {
+      C_1.push_back(point.x * e_1.x() + point.y * e_1.y());
+      C_2.push_back(point.x * e_2.x() + point.y * e_2.y());
+    }
+    float q = calcClosenessCriterion(C_1, C_2);
+    return -q;
+  };
+
+  int bits = 4;
```

Comment:
There is some decision regarding `bits` and `max_iter`:
- I chose `max_iter = 20` simply because when I ran my test without iteration limit, the algorithm would always converge in <15 iterations. I guess 20 is a good default, but it may depend on the environment and sensor kit.
- As for `bits=4`, I think it is slightly less precise than original algorithm (precision is 1 deg = 0.0174 rad). Float precision in interval [0, pi/2] is the worst in [1.0, pi/2]. The precision on the 4th bit is 0.0625, which is 3.58 degrees. The 1 degree precision is reached with the 6th bit (0.015625 = 0.90 deg). This can be verified [here](https://www.h-schmidt.net/FloatConverter/IEEE754.html) with number 0x3f820000. Of course my all reasoning assume the boost implementation uses `bits` the way I think. Still I think `bits=6` would be safer (3 degrees heading variation is big)